ravif/src/av1encoder.rs

```diff
@@ -6,6 +6,108 @@ use crate::rayoff as rayon;
+        MatrixCoefficients::YCgCo =>  avif_serialize::constants::MatrixCoefficients::Ycgco,
+        MatrixCoefficients::BT2020NCL =>  avif_serialize::constants::MatrixCoefficients::Bt2020Ncl,
+        MatrixCoefficients::BT2020CL =>  avif_serialize::constants::MatrixCoefficients::Bt2020Cl,
+        _ => avif_serialize::constants::MatrixCoefficients::Unspecified,
+    }
+}
+
+impl ColorSpace {
+    fn rgb_to_ycbcr<P: Pixel + Default>(self, bit_depth: BitDepth, px: Rgb<P>) -> [P;3] {
+        let depth = bit_depth.to_usize();
+        let max_value = ((1 << depth) - 1) as f32;
+        let rf = (u32::cast_from(px.r) as f32) / max_value;
+        let gf = (u32::cast_from(px.g) as f32) / max_value;
+        let bf = (u32::cast_from(px.b) as f32) / max_value;
+
+        let rgb_to_luma = match self {
+            ColorSpace::Srgb =>
+                [0.2126, 0.7152, 0.0722]
+            ,
+            ColorSpace::Rec2020Pq =>
+                [0.2627, 0.6780, 0.0593]
+            ,
+            ColorSpace::DisplayP3 =>
+                [0.22900385, 0.69172686, 0.07926947]
+            ,
+        };
+
+        let y = rgb_to_luma[0] * rf + rgb_to_luma[1]* gf + rgb_to_luma[2]* bf;
+        let cb = (bf - y) / (2.0*(1.0 - rgb_to_luma[2]));
+        let cr = (rf - y) / (2.0*(1.0 - rgb_to_luma[0]));
+
+        let y_int  = (y * max_value).round() as u16;
+        let cb_int = ((cb + 0.5)*max_value).round() as u16;
+        let cr_int = ((cr + 0.5)*max_value).round() as u16;
+
+        [P::cast_from(y_int), P::cast_from(cb_int), P::cast_from(cr_int)]
+    }
+
+    fn transfer_characteristics(self) -> TransferCharacteristics {
+            match self {
+                ColorSpace::Srgb => TransferCharacteristics::SRGB,
+                // For Rec.2020 there are a few valid transfer functions, two are currently supported here.
+                ColorSpace::Rec2020Pq => TransferCharacteristics::SMPTE2084,
+                // Display P3 uses the sRGB transfer function
+                ColorSpace::DisplayP3 => TransferCharacteristics::SRGB,
+            }
+        }
+    fn color_primaries(self) -> ColorPrimaries {
+        match self {
+            ColorSpace::Srgb => ColorPrimaries::BT709,
+            ColorSpace::Rec2020Pq => ColorPrimaries::BT2020,
+            ColorSpace::DisplayP3 => ColorPrimaries::SMPTE432
+        }
+    }
+
+    fn matrix_coefficients(self, color_model: ColorModel) ->  MatrixCoefficients {
+        match color_model{
+            ColorModel::YCbCr => match self {
+                ColorSpace::Srgb => MatrixCoefficients::BT709,
+                ColorSpace::Rec2020Pq => MatrixCoefficients::BT2020NCL,
+                // Since there's no matrix for Display P3 implemented, the one of BT709 is used.
+                // Although this isn't perfectly accurate, it should be acceptable as long as the decoder uses the same approach.
+                ColorSpace::DisplayP3 => MatrixCoefficients::BT709,
```

Comment:
Perhaps we should use the `BT709` matrix coefficients for `DisplayP3` inside `rgb_to_ycbcr()` in this case?

(at which point you can maybe create a `fn color_matrix(self) -> [3; f32]` on `MatrixCoefficients`?)

Reply:
Very good point! I used the exact RGB to luma conversion values for DisplayP3 with the matrix coefficients of Rec.709, this might have actually caused some bias. I'll double check this since I didn't notice this in test images.